File: backend/src/main/java/com/securechat/model/Message.java
package com.securechat.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import java.util.Date;
import java.util.Map;
import java.util.List;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "messages")
public class Message {
    @Id
    private String id;
    private String chatId;
    private String groupId;
    private String senderId;
    private String receiverId;
    private String encryptedMessage;
    // Maps userId -> encrypted AES key for that user
    private Map<String, String> encryptedKeys;
    private String mediaUrl;
    private String fileUrl;
    private String audioUrl;
    private String messageType; // text, image, video, file, voice
    private String status; // sent, delivered, read
    private List<Reaction> reactions;
    private Date timestamp = new Date();
}

File: backend_node/src/models/Message.js
const mongoose = require('mongoose');

const messageSchema = new mongoose.Schema({
  chatId: { type: mongoose.Schema.Types.ObjectId, ref: 'Chat', required: true, index: true },
  senderId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  receiverId: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  encryptedMessage: { type: String, required: true },
  encryptedKeys: { type: Map, of: String }, // userId -> encrypted AES key
  mediaUrl: { type: String },
  fileUrl: { type: String },
  audioUrl: { type: String },
  messageType: {
    type: String,
    enum: ['text', 'image', 'video', 'file', 'voice'],
    default: 'text',
  },
  status: {
    type: String,
    enum: ['sent', 'delivered', 'read'],
    default: 'sent',
  },
  reactions: [{
    userId: { type: String },
    emoji: { type: String },
  }],
  isDeleted: { type: Boolean, default: false },
}, {
  timestamps: true,
  toJSON: {
    transform: (doc, ret) => {
      ret.id = ret._id;
      ret.timestamp = ret.createdAt;
      delete ret.__v;
      return ret;
    }
  }
});

// Compound index for fetching messages in a chat sorted by time
messageSchema.index({ chatId: 1, createdAt: 1 });
messageSchema.index({ chatId: 1, status: 1 });

module.exports = mongoose.model('Message', messageSchema);
